### ISSUE TITLE
[Performance] [Tracking] Individual chat messages are slow to render

### ISSUE THREAD

**Issue by marcaaron** [Expensify/App#4022]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
- Tap on a chat

## Expected Result:
- When chat loads the entire visible area is filled with messages instantly

## Actual Result:
- When items first render only a subset of them render and each "batch" of items is slightly delayed and one message renders after the other in a staggered fashion

## Workaround:
Wait for the messages to appear awkwardly.

## Platform:
All platforms, but the effects are felt differently depending on the platform.

Here's an example of what this looks like on various platforms...

**Web:**

![2021-07-13_12-30-33 (1)](https://user-images.githubusercontent.com/32969087/125535102-446766ce-0911-44cc-af20-a50a791d1cb2.gif)

**iOS:**

![2021-07-13_12-58-21 (1)](https://user-images.githubusercontent.com/32969087/125535762-98a06c1a-df1c-4e89-ac9a-3637cc0afef4.gif)

**Version Number:**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @lschurr (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by lschurr** [comment]
@marcaaron should this be assigned to you?

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Luke9389 (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by lschurr** [comment]
Adding Eng before posting to External..

**Comment by Luke9389** [comment]
Ok, I see no reason not to make this external. Let's see what proposals our contributors come up with.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @SofiedeVreese (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by marcaaron** [comment]
@lschurr auto assign to eng is perfect thanks!

**Comment by SofiedeVreese** [comment]
Removing @marcaaron 's assignment.

Posted to Upwork: https://www.upwork.com/jobs/~011974a9e078cd37d9

Adding `Exported` label now.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @deetergp (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by aliabbasmalik8** [comment]
**PROPOSAL**

We can add `initialNumToRender` props in `FlatList` at this position

tested with `initialNumToRender={30}` and it's look good

https://github.com/Expensify/Expensify.cash/blob/1a1d66a322c733bbc79f8a5f05f1527240c769d7/src/components/InvertedFlatList/BaseInvertedFlatList.js#L142-L158


Thanks

**Comment by SofiedeVreese** [comment]
As I'm out of office tomorrow, I'm going to unassign myself and re-add the `External` label.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @kevinksullivan (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by SofiedeVreese** [comment]
Hey @kevinksullivan I'm OOO tomorrow hence applying the External label again.

The job is posted on Upwork!

**Comment by deetergp** [comment]
That sounds pretty straight-forward, @aliabbasmalik8, let's do it.

**Comment by marcaaron** [comment]
Sorry @deetergp I'm going to disagree here... more details incoming

**Comment by marcaaron** [comment]
The problem is that individual chat message items are slow to render. They'd be slow to render if we have 1 or 1000. So changing the `initialNumToRender` won't influence the total time it takes to load a chat message. It will only make it so that the FlatList will render with some minimum number of items.

That said, raising the `initialNumToRender` seems like a good idea to prevent the issue where blank space is observed. The default is 10 and seems like that is clearly what we are experiencing on iOS. I'll create a new issue so we can address that and make sure @aliabbasmalik8 is assigned.

**Comment by jsamr** [comment]
Might be related to
- #4101 

If you take a large picture of those commits, you'll notice a meta-pattern involving list cells, with 3 phases spanning over **2.3s**:

### Phase 1

(c33 → c52)
![image](https://user-images.githubusercontent.com/3646758/125946550-aca1edda-ef27-459a-bb59-948936087b7d.png)

`withOnyx(Component)` in `renderCell`. Consecutive 2-commits patterns caused by state changes (`preferredLocale` then `loading`) 

### Phase 2

(c52 → c71)
![image](https://user-images.githubusercontent.com/3646758/125946433-39f64c3c-31f7-4c10-a06d-70d5346f1ef7.png) 

`withOnyx(ReportActionItem)`, descendent of component in phase one.  Consecutive 2-commits patterns caused by state changes (`draftMessage` then `loading`)

### Phase 3

(c76 → c145)
![image](https://user-images.githubusercontent.com/3646758/125946246-12a4510d-e7d4-4272-84ce-1f2bc775ac31.png).
Two repeated components updates:
- 2 instances of `withOnyx(Component)`, descendant of component in phase 2.  Consecutive 2-commits patterns caused by state changes (`preferredLocale` then `loading`)
- 1 instance of `withOnyx(Component)` (the big spike of 30ms), which renders `ReportActionContextMenu`

### Discussion

It's hard to identify a clear pattern after phase 3, but it seems that it's when `RenderHtml` starts being rendered in each cell. The most important finding is that **bursts of commits identified in #4101 happen recursively**, for each component in the tree created by `renderCell` wrapped in a `withOnyx`, n + 1 extraneous commits are added.

So if we have, let's say, 3 components wrapped in `withOnyx` in the cell tree, each subscribing to two keys, that's 3 * (2 + 1) = **9 extraneous commits per cell**. Implementing the proposals suggested in #4101 should make a noticeable difference.

**Comment by marcaaron** [comment]
A simple observation (perhaps not surprising) but if we set the `initialNumToRender` for the `FlatList` to be very low + remove the loading spinner then the chat navigation happens very fast which seems to suggest that the time it takes to render items + the number rendered initially will also have a big impact (naturally we are rendering less of them - so if they are expensive to render then rendering a large number up front blocks us from navigating).

| Avg. Switch Time initialNumToRender={20} | initialNumToRender={1} | Difference  |
|----------------------------------|-------------------------------------------|-------------|
|                          1332.25 |                               282.5833333 | 1049.666667 |

Of course, in these examples, the chats are not "fully loaded" when initially rendered onscreen, but it "feels" faster vs. the interaction completely hanging for 1.5 seconds. If anything I think this just provides extra evidence that the individual items themselves are heavy and slow to render.

I'm starting to think that we're not looking for a silver bullet that will suddenly make everything faster, but lots of small changes to make report action items render faster + managing carefully how many of them we decide to render when the chat first loads.

I'm not too sure how to suggest we proceed, but will think about it. Might be interesting to set up a single test case and some profiling tools so that we can try to improve an "example" report action.

**Comment by quinthar** [comment]
To clarify, if you do this, will it only show one chat -- or will it start
showing one row, and then immediately backfill the rest?

**Comment by marcaaron** [comment]
Correct, it shows one row immediately. Then backfills the rest. Here's a video of what it looks like compared to what it looks like with the loading spinner + front-loading the report with 20 comments instead of 1

**Loading Spinner + initialNumToRender = 20**

https://user-images.githubusercontent.com/32969087/126252296-e68a0b60-b6a7-4a73-a891-42856153be67.mp4

**No loading spinner + initialNumToRender = 1**

https://user-images.githubusercontent.com/32969087/126252279-c94b0bf7-0172-4091-a146-f468ec3872b2.mp4

**Comment by quinthar** [comment]
Well that seems a huge improvement, right?

**Comment by marcaaron** [comment]
I think so, but the effect is slightly more noticeable with a larger chat... 

https://user-images.githubusercontent.com/32969087/126258234-486f10f5-e59b-4529-a2df-2086dc070c51.mp4

Mostly, I think there's likely a sweet spot where we:

1. make messages more performant to load (whether through refactors or Onyx improvements)
2. approximate an ideal number of messages to initially render based on the screen height (h/t to @rdjuric for that idea)

**Comment by quinthar** [comment]
I'd love to experience it with just one chat loading

**Comment by marcaaron** [comment]
https://user-images.githubusercontent.com/32969087/126264077-dfed5cd4-a4a1-459b-8c7c-c507a8f85fc7.mp4

**Comment by quinthar** [comment]
Oh sorry, I meant let's deploy this!

**Comment by MelvinBot** [comment]
@deetergp, @kevinksullivan Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by deetergp** [comment]
@marcaaron Just so I understand, you are suggesting that we:
* Get rid of the loading spinner
* Render the initial number of chats based on the vertical viewport height 
* Then finally backfill _n_ number of chats after that initial number loads?

Is that correct?

**Comment by mallenexpensify** [comment]
@marcaaron it appears you're working on this but you're not assigned.  Any reason @kevinksullivan shouldn't double this on Upwork? It's only $250 now and it's been 9 days

**Comment by marcaaron** [comment]
@mallenexpensify I'm going to make this more of a tracking issue and myself if you want to cancel the Upwork gig. This has become more of a discussion at this point and will need to be addressed by several smaller issues.

**Comment by marcaaron** [comment]
@deetergp yes, that's the idea @rdjuric had. I think maybe we can create a new issue for it.

**Comment by mallenexpensify** [comment]
Job is closed/removed in Upwork (also checked and `Help Wanted` label is on this issue)

**Comment by rdjuric** [comment]
> @deetergp yes, that's the idea @rdjuric had. I think maybe we can create a new issue for it.

@marcaaron I opened a issue for this https://github.com/Expensify/App/issues/4258

**Comment by marcaaron** [comment]
@rdjuric nice! Thanks for that.

**Comment by MelvinBot** [comment]
@marcaaron Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by marcaaron** [comment]
We're making some pretty interesting progress with this issue and it's starting to become clear that any list item components using `withOnyx` are going to take a long time to render since they'll always be blocked and "waiting for data" from Onyx. By leveraging the Context API we can avoid creating a new connection and get that time back. Exploring it in [this PR](https://github.com/Expensify/App/pull/4348).

**Comment by marcaaron** [comment]
So, I _think_ chat switching is starting to approach something resembling an acceptable amount of time. Check out these latest videos which are **release** builds off `main` + the changes in the last PR I linked above...

**iOS**

https://user-images.githubusercontent.com/32969087/127958695-935956fc-9268-4866-bef2-28fffd42577f.mov

**Android**

https://user-images.githubusercontent.com/32969087/127958741-0177a936-c80f-484c-99ee-37a580adc876.mp4

Android looks slightly slower than iOS still (and is a little janky or something at one part) but whatever we are doing is working. 🍻

**Comment by quinthar** [comment]
Whoa!  How do I get a taste of this?

**Comment by MelvinBot** [comment]
@marcaaron Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@marcaaron 6 days overdue. This is scarier than being forced to listen to Vogon poetry!

**Comment by MelvinBot** [comment]
@marcaaron Huh... This is 4 days overdue. Who can take care of this?

**Comment by MelvinBot** [comment]
@marcaaron Still overdue 6 days?! Let's take care of this!

**Comment by marcaaron** [comment]
Going to close this for now as the last of the changes are on staging. I'm not sure what else we can do to improve the individual loading of chat messages and they seem to load pretty fast for now.
